### PR TITLE
779 hash function

### PR DESCRIPTION
**Pull request by Arithmeticus**
First draft of hash function, proposed in #779.

Error message left as to-do item; guidance from editors appreciated.

I opted to leave out wrapper/cryptographic functionality, such as salting, and to demonstrate via example how it could be done by a developer on their own. In my opinion what we need here is a simple atomic function that can be incorporated into other molecular functions.

I may tinker with the prose description up to CG discussion, so comments are welcome.

### CODE REVIEW

File: specifications/xpath-functions-40/src/function-catalog.xml [hunk 1]
```diff
@@ -4758,7 +4758,124 @@ return normalize-unicode(concat($v1, $v2))</eg>
...
+               <fos:result>"9c9b913eb1b6254f4737ce947efd16f16e916f9d6ee5c1102a2002e48d4c88bd"</fos:result>
+            </fos:test>
+            <!--<fos:test>
+               <fos:expression>hash("password123" || $salt, "sha-1234567")</fos:expression>
+               <fos:error-result error-code="TODO: error code"/>
+            </fos:test>-->
+         </fos:example>
+      </fos:examples>
+      <fos:history>
+         <fos:version version="4.0">Newly proposed for 4.0.</fos:version>
```
Comment (ChristianGruen) [inline specifications/xpath-functions-40/src/function-catalog.xml:4875]:
```suggestion
         <fos:version version="4.0">New in 4.0.</fos:version>
```

…on the other hand, the document is not very consistent in this matter.

**Review by ChristianGruen — CHANGES_REQUESTED** [review]
Thanks, looks good otherwise.

File: specifications/xpath-functions-40/src/function-catalog.xml [hunk 2]
```diff
@@ -4758,7 +4758,124 @@ return normalize-unicode(concat($v1, $v2))</eg>
...
+            <code>fn:lower-case(fn:replace($function, '\W+', ''))</code>. </p>
+         <p>If <code>$value</code> is the empty sequence, or a string of zero length, the function
+            returns the empty sequence. </p>
+         <p>If <code>$value</code> is an instance of 
+            <code>xs:string</code> it is cast to <code>xs:hexBinary</code> on the basis of UTF-8
+            encoding. If <code>$value</code> is an instance of 
+            <code>xs:base64Binary</code> it is cast to <code>xs:hexBinary</code>. </p>
+         <p>The function returns an <code>xs:string</code> representation of the bytes returned by
+            passing the <code>xs:hexBinary</code> value of <code>$value</code> as an octet sequence
+            through the specified hash or checksum function. Output is always lowercase.</p>
```
[the remark this answers is not in the document]
Comment (Arithmeticus) [reply specifications/xpath-functions-40/src/function-catalog.xml:4803]: For lowercase, I was taking my lead from common usage/expectations as well as the specs for the MD5 and SHA-1 algorithms, which prefer lowercase.

**Comment by Arithmeticus** [comment]
Thanks, Christian, for the excellent suggestions and catches. Latest commit addresses them all, + a few more. To-do / desiderata would be examples using binary input.

File: specifications/xpath-functions-40/src/function-catalog.xml [hunk 3]
```diff
@@ -4758,7 +4758,124 @@ return normalize-unicode(concat($v1, $v2))</eg>
...
+         <p>Returns a string representation of the results from a specified hash, checksum, or
+            cyclic redundancy check function upon the input.</p>
+      </fos:summary>
+      <fos:rules>
+         <p>If the zero-argument version of the function is used, the result is the same as calling
+            the one-argument version, with <code>$value</code> set to <code>fn:string(.)</code>. </p>
+         <p>If the one-argument version of the function is used, the result is the same as calling
+            the two-argument version, with <code>$algorithm</code> set to "md5". </p>
+         <p>The effective value of <code>$algorithm</code> is the value of the expression
+            <code>fn:lower-case(fn:replace($function, '\W+', ''))</code>. </p>
```
[the remark this answers is not in the document]
Comment (Arithmeticus) [reply specifications/xpath-functions-40/src/function-catalog.xml:4794]: I was looking for a way to forgive a user if they couldn't remember if a name had a hyphen or not, but my solution was too lenient.

File: specifications/xpath-functions-40/src/function-catalog.xml [hunk 4]
```diff
@@ -4758,7 +4758,122 @@ return normalize-unicode(concat($v1, $v2))</eg>
          </fos:example>
       </fos:examples>
    </fos:function>
-
+   <fos:function name="hash" prefix="fn" at="2024-01-10" diff="add">
+      <fos:signatures>
+         <fos:proto name="hash" return-type="xs:string">
+            <fos:arg name="value" type="union(xs:string, xs:hexBinary, xs:base64Binary)"
+               default="fn:string(.)"/>
```
Comment (ChristianGruen) [inline specifications/xpath-functions-40/src/function-catalog.xml:4765]: The default for the first argument should be dropped.

File: specifications/xpath-functions-40/src/function-catalog.xml [hunk 5]
```diff
@@ -4758,7 +4758,122 @@ return normalize-unicode(concat($v1, $v2))</eg>
...
+      </fos:summary>
+      <fos:rules>
+         <p>If the one-argument version of the function is used, the result is the same as calling
+            the two-argument version, with <code>$algorithm</code> set to "MD5". </p>
+         <p>The effective value of <code>$algorithm</code> is the value of the expression
+            <code>fn:upper-case(fn:normalize-space($algorithm))</code>. </p>
+         <p>If <code>$value</code> is the empty sequence, the function returns the empty sequence.</p>
+         <p>If <code>$value</code> is an instance of <code>xs:string</code> it is cast to a sequence
+            of octets on the basis of UTF-8 encoding. If <code>$value</code> is an instance of
+               <code>xs:base64Binary</code> or <code>xs:hexBinary</code> it is cast to a sequence of
```
Comment (ChristianGruen) [inline specifications/xpath-functions-40/src/function-catalog.xml:4791]: Maybe “converted” or “translated”? “cast” may imply that “sequence of octets” is a data type.

File: specifications/xpath-functions-40/src/function-catalog.xml [hunk 6]
```diff
@@ -4758,7 +4758,122 @@ return normalize-unicode(concat($v1, $v2))</eg>
...
+         <p>Returns a string representation of the results from a specified hash, checksum, or
+            cyclic redundancy check function upon the input.</p>
+      </fos:summary>
+      <fos:rules>
+         <p>If the one-argument version of the function is used, the result is the same as calling
+            the two-argument version, with <code>$algorithm</code> set to "MD5". </p>
+         <p>The effective value of <code>$algorithm</code> is the value of the expression
+            <code>fn:upper-case(fn:normalize-space($algorithm))</code>. </p>
+         <p>If <code>$value</code> is the empty sequence, the function returns the empty sequence.</p>
+         <p>If <code>$value</code> is an instance of <code>xs:string</code> it is cast to a sequence
```
Comment (ChristianGruen) [inline specifications/xpath-functions-40/src/function-catalog.xml:4789]:
```suggestion
         <p>If <code>$value</code> is an instance of <code>xs:string</code>, it is cast to a sequence
```

**Review by ChristianGruen — CHANGES_REQUESTED** [review]
…some more observations, all minor.

File: specifications/xpath-functions-40/src/function-catalog.xml [hunk 7]
```diff
@@ -4758,7 +4758,121 @@ return normalize-unicode(concat($v1, $v2))</eg>
          </fos:example>
       </fos:examples>
    </fos:function>
-
+   <fos:function name="hash" prefix="fn" at="2024-01-10" diff="add">
+      <fos:signatures>
+         <fos:proto name="hash" return-type="xs:string">
+            <fos:arg name="value" type="union(xs:string, xs:hexBinary, xs:base64Binary)"/>
```
Comment (ChristianGruen) [inline specifications/xpath-functions-40/src/function-catalog.xml:4764]:
```suggestion
            <fos:arg name="value" type="union(xs:string, xs:hexBinary, xs:base64Binary)?"/>
```

**Comment by benibela** [comment]
I still do not understand why this would be put into fn and not into a separate hash module. Or it could fit into Expath's binary module

md5 and sha-1 are already deprecated anyways

**Comment by ndw** [comment]
The CG agreed to merge this PR at [meeting 062](https://qt4cg.org/meeting/minutes/2024/01-23.html)

**Pull request merged**